src/diagnostics/lintstore.md

```diff
@@ -0,0 +1,40 @@
+# LintStore
+
+This page documents some of the machinery around lint registration and how we
+run lints in the compiler.
+
+The `LintStore` is the central piece of infrastructure, around which everything
+rotates. It's not available during the early parts of compilation (i.e., before
+TyCtxt) in most of the code, as we need to fill it in with all of the lints,
+which can only happen after plugin registration.
+
+We store essentially two separate things in the `LintStore`: a list of all known
+lints with associated metadata (default lint levels, notably), and the lint
+passes. We store constructors for the lint passes rather than the passes
+themselves, because lint passes often want to change state to track something
+in the code (so their methods take `&mut self`) and we want to avoid needing to
+synchronize access to the LintStore.
+
+For the most part, we currently only instantiate each lint pass once, though in
+the future we'll likely want more fine-grained linting to better support
+incremental compilation. We already have late module passes which are
+constructed per-module if the module changes in an incremental session.
+
+## Registration
+
+We must register both every lint and every lint pass into the `LintStore` for
+the code to work properly. This often is as simple as calling
+`register_lints(PassType::get_lints())` and then `register_late_pass(||
+Box::new(PassType))`.
+
+Within the compiler, for performance reasons, we usually do not register dozens
```

Comment:
This is a good detail to include. I would add a section header before it like

\#\# Compiler lint passes are combined into one pass